service/policy/db/migrations/20241125220354_keys_table.sql

```diff
@@ -0,0 +1,233 @@
+        key_id uuid NOT NULL REFERENCES public_keys (id) ON DELETE CASCADE,
+        PRIMARY KEY (definition_id, key_id)
+    );
+
+CREATE TABLE IF NOT EXISTS
+    attribute_value_public_key_map (
+        value_id uuid NOT NULL REFERENCES attribute_values (id) ON DELETE CASCADE,
+        key_id uuid NOT NULL REFERENCES public_keys (id) ON DELETE CASCADE,
+        PRIMARY KEY (value_id, key_id)
+    );
+
+-- Trigger function to update was_used column
+CREATE
+OR REPLACE FUNCTION update_was_used () RETURNS trigger AS $$
+BEGIN
+    UPDATE public_keys SET was_used = TRUE WHERE id = NEW.key_id;
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Trigger for attribute_namespace_key_map
+CREATE TRIGGER trigger_update_was_used_namespace
+AFTER INSERT ON attribute_namespace_public_key_map FOR EACH ROW
+EXECUTE FUNCTION update_was_used ();
+
+-- View for active namespace keys
+CREATE VIEW
+    active_namespace_public_keys_view AS
+SELECT
+    km.namespace_id,
+    jsonb_agg(
+        DISTINCT jsonb_build_object(
+            'id',
+            ky.id,
+            'is_active',
+            ky.is_active,
+            'was_used',
+            ky.was_used,
+            'public_key',
+            json_build_object(
+                'alg',
+                ky.alg,
+                'kid',
+                ky.key_id,
+                'pem',
+                ky.public_key
+            ),
+            'kas',
+            jsonb_build_object('id', kas.id, 'uri', kas.uri, 'name', kas.name)
+        )
+    ) AS keys
+FROM
+    public_keys AS ky
+    INNER JOIN attribute_namespace_public_key_map AS km ON ky.id = km.key_id
+    LEFT JOIN key_access_servers AS kas ON ky.key_access_server_id = kas.id
+WHERE
+    ky.is_active = TRUE
+GROUP BY
+    km.namespace_id;
+
+-- Trigger for attribute_definition_key_map
+CREATE TRIGGER trigger_update_was_used_definition
+AFTER INSERT ON attribute_definition_public_key_map FOR EACH ROW
+EXECUTE FUNCTION update_was_used ();
+
+-- View for active definition keys
+CREATE VIEW
+    active_definition_public_keys_view AS
+SELECT
+    km.definition_id,
+    jsonb_agg(
+        DISTINCT jsonb_build_object(
+            'id',
+            ky.id,
+            'is_active',
+            ky.is_active,
+            'was_used',
+            ky.was_used,
+            'public_key',
+            json_build_object(
+                'alg',
+                ky.alg,
+                'kid',
+                ky.key_id,
+                'pem',
+                ky.public_key
+            ),
+            'kas',
+            jsonb_build_object('id', kas.id, 'uri', kas.uri, 'name', kas.name)
+        )
+    ) AS keys
+FROM
+    public_keys AS ky
+    INNER JOIN attribute_definition_public_key_map AS km ON ky.id = km.key_id
+    LEFT JOIN key_access_servers AS kas ON ky.key_access_server_id = kas.id
+WHERE
+    ky.is_active = TRUE
+GROUP BY
+    km.definition_id;
+
+-- Trigger for attribute_value_key_map
+CREATE TRIGGER trigger_update_was_used_value
```

Comment:
Can you please help me understand the tradeoff you chose keeping this permanent state in sync with a column and transactions instead of doing the JOINs as needed when checking if a key is assigned to a policy object?